cmd/cli/plugin/cluster/upgrade.go

```diff
@@ -333,3 +349,65 @@ func getClusterTKRNameFromClusterLabels(clusterLabels map[string]string) (string
 	}
 	return tkrName, nil
 }
+
+func getValidTKRVersionFromClusterForUpgrade(cluster *capiv1.Cluster, tkrName string) (string, error) {
+	tkrVersion := strings.ReplaceAll(tkrName, "---", "+")
+	// If existing TKR on the cluster and the one requested are same
+	// continue the upgrade using the given TKR
+	if clusterTKR, exists := cluster.Labels[runv1alpha3.LabelTKR]; exists && clusterTKR == tkrName {
+		return tkrVersion, nil
+	}
+
+	updates, err := topology.GetClusterAvailableUpgrades(cluster)
+	if err != nil {
+		return "", err
+	}
+
+	matchingUpdates := updates.Filter(func(tv string) bool {
+		if strings.HasPrefix(tv, tkrVersion) {
+			return true
+		}
+		return false
+	})
+	if len(matchingUpdates) == 0 {
+		updateNames := updates.Map(func(tv string) string {
+			return strings.ReplaceAll(tv, "+", "---")
+		})
+		return "", errors.Errorf("cluster cannot be upgraded, no compatible upgrades found matching the TKR name/prefix '%s', available compatible upgrades %s", tkrName, updateNames.Keys())
+	}
+	return latestTkrVersion(matchingUpdates.Keys()), nil
+}
+
+func getClusterResource(clusterClient clusterclient.Client, clusterName, clusterNamespace string) (*capiv1.Cluster, error) {
+	if clusterNamespace == "" {
+		clusterNamespace = constants.DefaultNamespace
+	}
+	cluster := &capiv1.Cluster{}
+	err := clusterClient.GetResource(cluster, clusterName, clusterNamespace, nil, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get cluster %q from namespace %q", clusterName, clusterNamespace)
+	}
+	return cluster, nil
+}
+
+func clusterHasUpdatesAvailableCondition(cluster *capiv1.Cluster) bool {
+	for _, condition := range cluster.Status.Conditions {
+		if condition.Type == runv1alpha3.ConditionUpdatesAvailable {
+			return true
+		}
+	}
+	return false
```

Comment:
```suggestion
	return capiconditions.Has(cluster, runv1alpha3.ConditionUpdatesAvailable)
```

Perhaps, this method is not even necessary.